kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```
@@ -112,65 +135,121 @@ else if (requestFilterResult.message() != null) {
-        if (msg instanceof DecodedResponseFrame<?> decodedFrame) {
-            if (decodedFrame instanceof InternalResponseFrame<?> frame && frame.isRecipient(filter)) {
+        if (msg instanceof InternalResponseFrame<?> decodedFrame) {
+            // jump the queue, let extra requests flow back to their sender
+            if (decodedFrame.isRecipient(filter)) {
                 if (LOGGER.isDebugEnabled()) {
                     LOGGER.debug("{}: Completing {} response for request sent by this filter{}: {}",
                             ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
                 }
-                CompletableFuture<ApiMessage> p = frame.promise();
+                CompletableFuture<ApiMessage> p = decodedFrame.promise();
                 p.complete(decodedFrame.body());
             }
             else {
-                var filterContext = new DefaultFilterContext(filter, ctx, decodedFrame, null, timeoutMs, sniHostname);
-                if (LOGGER.isDebugEnabled()) {
-                    LOGGER.debug("{}: Dispatching upstream {} response to filter {}: {}",
-                            ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
-                }
-                var stage = invoker.onResponse(decodedFrame.apiKey(), decodedFrame.apiVersion(),
-                        decodedFrame.header(), decodedFrame.body(), filterContext);
-                if (stage == null) {
-                    LOGGER.error("{}: Filter{} for {} response unexpectedly returned null. This is a coding error in the filter. Closing connection.",
-                            ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
+                doRead(ctx, msg);
+            }
+        }
+        else if (readFuture.isDone()) {
+            readFuture = doRead(ctx, msg);
+        }
+        else {
+            readFuture = readFuture.whenComplete((a, b) -> doRead(ctx, msg));
```

are we swallowing an error (case where b != null)

---

I've added unit tests and push fixes to my branch.